File: app/src/main/java/com/example/dualaudioplayer/AudioListAdapter.java
package com.example.dualaudioplayer;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;

public class AudioListAdapter extends RecyclerView.Adapter<AudioListAdapter.ViewHolder> {
    private final List<AudioItem> audioList;
    private final OnItemClickListener listener;

    interface OnItemClickListener { void onItemClick(AudioItem item); }

    AudioListAdapter(List<AudioItem> audioList, OnItemClickListener listener) {
        this.audioList = audioList;
        this.listener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_audio, parent, false));
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bind(audioList.get(position), listener);
    }

    @Override
    public int getItemCount() { return audioList.size(); }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        final TextView titleTextView; final TextView artistTextView;
        ViewHolder(View itemView) {
            super(itemView);
            titleTextView = itemView.findViewById(R.id.titleTextView);
            artistTextView = itemView.findViewById(R.id.artistTextView);
        }
        void bind(final AudioItem item, final OnItemClickListener listener) {
            titleTextView.setText(item.title);
            artistTextView.setText(item.artist);
            itemView.setOnClickListener(v -> listener.onItemClick(item));
        }
    }
}

File: app/src/main/kotlin/com/example/dualaudioplayer/AudioListAdapter.kt
package com.example.dualaudioplayer
import android.view.LayoutInflater; import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil; import androidx.recyclerview.widget.ListAdapter; import androidx.recyclerview.widget.RecyclerView
import coil.load
import com.example.dualaudioplayer.databinding.ListItemAudioBinding
class AudioListAdapter(private val onItemClicked: (AudioItem) -> Unit) : ListAdapter<AudioItem, AudioListAdapter.AudioViewHolder>(AudioDiffCallback) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = AudioViewHolder(ListItemAudioBinding.inflate(LayoutInflater.from(parent.context), parent, false))
    override fun onBindViewHolder(holder: AudioViewHolder, position: Int) = holder.bind(getItem(position), onItemClicked)
    class AudioViewHolder(private val binding: ListItemAudioBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(item: AudioItem, onItemClicked: (AudioItem) -> Unit) {
            binding.titleTextView.text = item.title; binding.artistTextView.text = item.artist
            binding.albumArtImageView.load(item.albumArtUri) { placeholder(android.R.drawable.ic_media_play); error(android.R.drawable.ic_media_play) }
            itemView.setOnClickListener { onItemClicked(item) }
        }
    }
    object AudioDiffCallback : DiffUtil.ItemCallback<AudioItem>() {
        override fun areItemsTheSame(oldItem: AudioItem, newItem: AudioItem): Boolean = oldItem.uri == newItem.uri
        override fun areContentsTheSame(oldItem: AudioItem, newItem: AudioItem): Boolean = oldItem == newItem
    }
}
